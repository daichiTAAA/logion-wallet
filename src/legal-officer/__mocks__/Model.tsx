import {
    fetchRecoveryInfo,
} from './ModelMock';

export {
    fetchRecoveryInfo,
};
