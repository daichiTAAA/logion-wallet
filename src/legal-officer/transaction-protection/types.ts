import { UUID } from "../../logion-chain/UUID"

export type LocItemStatus = 'DRAFT' | 'PUBLISHED'

export type LocItemType = 'Data' | 'Document' | 'Linked LOC'

export interface LocItem {
    name: string,
    value: string,
    timestamp: string | null,
    type: LocItemType,
    submitter: string,
    status: LocItemStatus,
    nature?: string,
    target?: UUID,
    newItem: boolean,
}

export enum PublishStatus {
    NONE,
    START,
    PUBLISH_PENDING,
    PUBLISHING,
    PUBLISHED
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem
}
