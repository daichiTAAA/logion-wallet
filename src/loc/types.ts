import { SignAndSubmit } from "../ExtrinsicSubmitter"
import { UUID } from "../logion-chain/UUID"

export type LocItemStatus = 'DRAFT' | 'PUBLISHED'

export type LocItemType = 'Data' | 'Document' | 'Linked LOC'

export interface LocItem {
    name: string,
    value: string,
    timestamp: string | null,
    type: LocItemType,
    submitter: string,
    status: LocItemStatus,
    nature?: string,
    target?: UUID,
    newItem: boolean,
    linkDetailsPath?: string,
}

export enum PublishStatus {
    NONE,
    START,
    PUBLISH_PENDING,
    PUBLISHING,
    PUBLISHED,
    ERROR
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem;
    itemType: 'Public Data' | 'Document' | 'Link';
    signAndSubmitFactory: ((item: LocItem) => SignAndSubmit) | null;
    confirm: ((item: LocItem) => void) | null;
}

export type Viewer = 'User' | 'LegalOfficer';
