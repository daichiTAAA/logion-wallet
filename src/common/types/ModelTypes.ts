import { LocType, MetadataItem, Link, CollectionItem } from '../../logion-chain/Types';
import Identity from './Identity';
import PostalAddress from './PostalAddress';
import { File } from '../../logion-chain/Types'
import UserIdentity from './Identity';

export type ProtectionRequestStatus = "PENDING" | "REJECTED" | "ACCEPTED" | "ACTIVATED";

export interface LegalOfficerDecision {
    rejectReason: string | null,
    decisionOn: string | null,
    locId?: string | null,
}

export interface ProtectionRequest {
    id: string,
    requesterAddress: string,
    decision: LegalOfficerDecision,
    userIdentity: Identity,
    userPostalAddress: PostalAddress,
    createdOn: string,
    isRecovery: boolean,
    addressToRecover: string | null,
    status: ProtectionRequestStatus,
    legalOfficerAddress: string,
    otherLegalOfficerAddress: string,
}

export interface Transaction {
    from: string,
    to: string,
    pallet: string,
    method: string,
    transferValue: string,
    tip: string,
    fee: string,
    reserved: string,
    total: string,
    createdOn: string,
    type: string,
    successful: boolean,
    error?: TransactionError
}

export interface TransactionError {
    section: string,
    name: string,
    details: string
}

export interface TransactionsSet {
    transactions: Transaction[],
}

export interface AddedOn {
    addedOn: string;
}

/**
 * Blockchain File, extended with private attributes and timestamp.
 */
export interface LocFile extends File, AddedOn {
    name: string;
}

/**
 * Blockchain MetadataItem, extended with timestamp.
 */
export interface LocMetadataItem extends MetadataItem, AddedOn {
}

/**
 * Blockchain MetadataItem, extended with timestamp.
 */
export interface LocLink extends Link, AddedOn {
    target: string; // is redundant with inherited "id: UUID"
}

export interface LocCollectionItem extends AddedOn {
    itemId: string,
}

export type MergedCollectionItem = CollectionItem & AddedOn

export interface LocRequestVoidInfo {
    reason?: string; // undefined in public view
    voidedOn?: string;
}

export interface LocRequest {
    ownerAddress: string;
    requesterAddress?: string | null;
    requesterIdentityLoc?: string | null;
    description: string;
    locType: LocType;
    createdOn: string;
    decisionOn?: string;
    id: string;
    status: "OPEN" | "REQUESTED" | "REJECTED" | "CLOSED";
    rejectReason?: string;
    userIdentity?: Identity;
    closedOn?: string;
    files: LocFile[];
    metadata: LocMetadataItem[];
    links: LocLink[];
    voidInfo?: LocRequestVoidInfo;
}

export type LocRequestStatus = "OPEN" | "REQUESTED" | "REJECTED" | "CLOSED";

export interface LocRequestFragment {
    requesterAddress?: string | null;
    requesterIdentityLoc?: string | null;
    locType: LocType;
    userIdentity?: UserIdentity;
}

export function isLogionIdentityLoc(loc: LocRequestFragment): boolean {
    return loc.locType === 'Identity' && !loc.requesterAddress && !loc.requesterIdentityLoc;
}

export function isLogionTransactionLoc(loc: LocRequestFragment): boolean {
    return loc.locType === 'Transaction' && (loc.requesterIdentityLoc !== undefined && loc.requesterIdentityLoc !== null);
}
