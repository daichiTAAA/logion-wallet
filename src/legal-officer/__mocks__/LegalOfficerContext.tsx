export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let acceptedTokenizationRequests: any[] | null = null;

export let rejectedTokenizationRequests: any[] | null = null;

export let refreshRequests = () => {

};

export let pendingProtectionRequests: any[] | null = null;

export let acceptedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export function useLegalOfficerContext() {
    return {
        pendingTokenizationRequests,
        rejectRequest,
        rejectedTokenizationRequests,
        acceptedTokenizationRequests,
        refreshRequests,
        pendingProtectionRequests,
        acceptedProtectionRequests,
        protectionRequestsHistory,
    };
}

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setRejectedRequests(requests: any[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: any[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setAcceptedProtectionRequests(requests: any[]) {
    acceptedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}
