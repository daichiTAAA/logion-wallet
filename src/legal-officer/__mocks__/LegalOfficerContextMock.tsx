export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let acceptedTokenizationRequests: any[] | null = null;

export let rejectedTokenizationRequests: any[] | null = null;

export let refreshRequests = jest.fn();

export let pendingProtectionRequests: any[] | null = null;

export let activatedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export let recoveryRequestsHistory: any[] | null = null;

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setRejectedRequests(requests: any[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: any[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setActivatedProtectionRequests(requests: any[]) {
    activatedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}

export function setRecoveryRequestsHistory(requests: any[]) {
    recoveryRequestsHistory = requests;
}
