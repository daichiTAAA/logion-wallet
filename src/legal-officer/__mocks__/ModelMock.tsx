export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob

export let rejectRequest = jest.fn();

export let acceptRequest = jest.fn();

export function setAcceptRequest(mockFn: any) {
    acceptRequest = mockFn;
}

export let setAssetDescription = jest.fn();

export let rejectProtectionRequest = jest.fn();

export function setRejectProtectionRequest(mockFn: any) {
    rejectProtectionRequest = mockFn;
}

export let acceptProtectionRequest = jest.fn();

export function setAcceptProtectionRequest(mockFn: any) {
    acceptProtectionRequest = mockFn;
}

export function decision(legalOfficerAddress: string, decisions: any[]) {
    return decisions[0];
}

export let fetchRecoveryInfo = jest.fn();

export function setFetchRecoveryInfo(mockFn: any) {
    fetchRecoveryInfo = mockFn;
}
