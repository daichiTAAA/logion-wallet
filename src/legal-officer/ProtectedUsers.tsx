import React from 'react';
import { Table } from "react-bootstrap";
import { useLegalOfficerContext } from "./LegalOfficerContext";
import UserInfo from "../common/UserInfo";
import LegalOfficerInfo from "../common/LegalOfficerInfo";
import { useCommonContext } from "../common/CommonContext";
import { legalOfficerByAddress } from "../common/types/LegalOfficer";

export default function ProtectedUsers() {
    const { activatedProtectionRequests } = useLegalOfficerContext();
    const { currentAddress } = useCommonContext();

    return (
            <>
                <Table striped bordered responsive>
                    <thead>
                    <tr>
                        <th>User</th>
                        <th>Other Legal Officer</th>
                    </tr>
                    </thead>
                    <tbody>
                    {
                        activatedProtectionRequests?.map(request => {
                            const otherLegalOfficerAddress = request.decisions
                                    .map(decision => decision.legalOfficerAddress)
                                    .find(legalOfficerAddress => legalOfficerAddress !== currentAddress);
                            return (
                                    <tr key={request.id}>
                                        <td>
                                            <UserInfo
                                                    address={request.requesterAddress}
                                                    userIdentity={request.userIdentity}
                                                    postalAddress={request.userPostalAddress}/>
                                        </td>
                                        <td>
                                            {otherLegalOfficerAddress !== undefined &&
                                            <LegalOfficerInfo
                                                    legalOfficer={legalOfficerByAddress(otherLegalOfficerAddress)}/>
                                            }
                                        </td>
                                    </tr>
                            )
                        })
                    }
                    </tbody>
                </Table>
            </>
    );
}
