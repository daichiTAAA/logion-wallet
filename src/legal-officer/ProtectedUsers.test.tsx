jest.mock("../common/RootContext");
jest.mock('./LegalOfficerContext');

import { ProtectionRequest } from "../common/types/ModelTypes";
import React from 'react';
import ProtectedUsers from "./ProtectedUsers";
import { shallowRender } from '../tests';
import moment from "moment";
import { setActivatedProtectionRequests } from "./__mocks__/LegalOfficerContextMock";

test('renders one row', () => {
    setActivatedProtectionRequests([
        request('2eb5f71c-7f31-44b5-9390-c3bf56501880')
    ]);
    const tree = shallowRender(<ProtectedUsers />);
    expect(tree).toMatchSnapshot();
});

test('renders 2 rows', () => {
    setActivatedProtectionRequests([
        request('2eb5f71c-7f31-44b5-9390-c3bf56501880'),
        request('9296d068-613f-4b53-9d45-d9cd494d2a45')
    ]);
    const tree = shallowRender(<ProtectedUsers />);
    expect(tree).toMatchSnapshot();
});

function request(id: string): ProtectionRequest {
    return  {
        id,
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        decisions: [
            {
                legalOfficerAddress: '5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty',
                status: 'ACCEPTED',
                rejectReason: null,
                decisionOn: null
            },
            {
                legalOfficerAddress: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY',
                status: 'ACCEPTED',
                rejectReason: null,
                decisionOn: null
            }
        ],
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        isRecovery: false,
        addressToRecover: "",
        status: "PENDING",
        createdOn: moment('2021-06-10T13:48:00.000Z').toISOString(),
    };
}
