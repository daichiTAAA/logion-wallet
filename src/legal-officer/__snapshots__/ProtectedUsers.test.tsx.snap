// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 2 rows 1`] = `
<React.Fragment>
  <ForwardRef
    bordered={true}
    responsive={true}
    striped={true}
  >
    <thead>
      <tr>
        <th>
          User
        </th>
        <th>
          Other Legal Officer
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <UserInfo
            address="5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW"
            postalAddress={
              Object {
                "city": "Liège",
                "country": "Belgium",
                "line1": "Place de le République Française, 10",
                "line2": "boite 15",
                "postalCode": "4000",
              }
            }
            userIdentity={
              Object {
                "email": "john.doe@logion.network",
                "firstName": "John",
                "lastName": "Doe",
                "phoneNumber": "+1234",
              }
            }
          />
        </td>
        <td>
          <LegalOfficerInfo
            legalOfficer={
              Object {
                "address": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
                "details": "SELARL ADRASTEE
Gare des Brotteaux
14, place Jules Ferry
69006 LYON
France
Tel: +33 4 78 52 87 56",
                "email": "g.grain@adrastee-lyon.fr",
                "name": "Guillaume Grain",
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          <UserInfo
            address="5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW"
            postalAddress={
              Object {
                "city": "Liège",
                "country": "Belgium",
                "line1": "Place de le République Française, 10",
                "line2": "boite 15",
                "postalCode": "4000",
              }
            }
            userIdentity={
              Object {
                "email": "john.doe@logion.network",
                "firstName": "John",
                "lastName": "Doe",
                "phoneNumber": "+1234",
              }
            }
          />
        </td>
        <td>
          <LegalOfficerInfo
            legalOfficer={
              Object {
                "address": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
                "details": "SELARL ADRASTEE
Gare des Brotteaux
14, place Jules Ferry
69006 LYON
France
Tel: +33 4 78 52 87 56",
                "email": "g.grain@adrastee-lyon.fr",
                "name": "Guillaume Grain",
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </ForwardRef>
</React.Fragment>
`;

exports[`renders one row 1`] = `
<React.Fragment>
  <ForwardRef
    bordered={true}
    responsive={true}
    striped={true}
  >
    <thead>
      <tr>
        <th>
          User
        </th>
        <th>
          Other Legal Officer
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <UserInfo
            address="5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW"
            postalAddress={
              Object {
                "city": "Liège",
                "country": "Belgium",
                "line1": "Place de le République Française, 10",
                "line2": "boite 15",
                "postalCode": "4000",
              }
            }
            userIdentity={
              Object {
                "email": "john.doe@logion.network",
                "firstName": "John",
                "lastName": "Doe",
                "phoneNumber": "+1234",
              }
            }
          />
        </td>
        <td>
          <LegalOfficerInfo
            legalOfficer={
              Object {
                "address": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
                "details": "SELARL ADRASTEE
Gare des Brotteaux
14, place Jules Ferry
69006 LYON
France
Tel: +33 4 78 52 87 56",
                "email": "g.grain@adrastee-lyon.fr",
                "name": "Guillaume Grain",
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </ForwardRef>
</React.Fragment>
`;
